### DISCUSSION TITLE
Intel compiler errors for install.

### DISCUSSION THREAD

**Discussion by HathewayWill** [dtcenter/METplus#1962] [Installation] [answered]
This is my first time trying to get the intel compilers to work with MET & METPLUS.

Even though MET says that it has installed I noticed that not all the /bin files are created.

Here are my logs and my environment set up.

`

export PYTHON_VERSION=$(/usr/bin/python3 -V 2>&1|awk '{print $2}')
export PYTHON_VERSION_MAJOR_VERSION=$(echo $PYTHON_VERSION | awk -F. '{print $1}')
export PYTHON_VERSION_MINOR_VERSION=$(echo $PYTHON_VERSION | awk -F. '{print $2}')
export PYTHON_VERSION_COMBINED=$PYTHON_VERSION_MAJOR_VERSION.$PYTHON_VERSION_MINOR_VERSION

export CC=/opt/intel/oneapi/compiler/2022.2.1/linux/bin/intel64/icc
export CXX=/opt/intel/oneapi/compiler/2022.2.1/linux/bin/intel64/icpc
export FC=/opt/intel/oneapi/compiler/2022.2.1/linux/bin/intel64/ifort
export F77=/opt/intel/oneapi/compiler/2022.2.1/linux/bin/intel64/ifort
export F90=/opt/intel/oneapi/compiler/2022.2.1/linux/bin/intel64/ifort
export gcc_version=$(icc -dumpversion -diag-disable=10441)
export TEST_BASE=$WRF_FOLDER/MET-10.1.2
export COMPILER=intel_$gcc_version
export MET_SUBDIR=${TEST_BASE}
export MET_TARBALL=met-10.1.2.20220516.tar.gz
export USE_MODULES=FALSE
export MET_PYTHON=/usr
export MET_PYTHON_CC=-I${MET_PYTHON}/include/python${PYTHON_VERSION_COMBINED}
export MET_PYTHON_LD=-L${MET_PYTHON}/lib/python${PYTHON_VERSION_COMBINED}/config-${PYTHON_VERSION_COMBINED}-x86_64-linux-gnu\ -L${MET_PYTHON}/lib\ -lpython${PYTHON_VERSION_COMBINED}\ -lcrypt\ -lpthread\ -ldl\ -lutil\ -lm
export SET_D64BIT=FALSE

chmod 775 compile_MET_all.sh
./compile_MET_all.sh`

[met_metplus_logs.zip](https://github.com/dtcenter/METplus/files/10035612/met_metplus_logs.zip)

**Comment by HathewayWill** [comment]
scripts for running met and metplus with intel.

[met_metplus_scripts.zip](https://github.com/dtcenter/METplus/files/10043665/met_metplus_scripts.zip)

**Comment by HathewayWill** [comment]
interesting development

if I remove the -libutil form the MET_PYTHON_LD the met and metplus compilation works.  What exactly is -libutil do?

**Reply by jprestop** [reply]
Hi @HathewayWill.  Unfortunately, I am not incredibly familiar with the various libraries that Python needs to link with. I did a Google search for libutil.so and found various things, but none of which I am confident or comfortable with passing along saying definitely what libutil does. 

If you run "python3-config --ldflags", does the results contain "-lutil"?  If so, it seems MET_PYTHON_LD would need "-lutil".  If not, then it seems MET_PYTHON_LD would not need "-lutil".  We know from experience that running "python3-config --ldflags" does not always contain all of the necessary flags, but we usually don't need to remove any, unlike running "python3-config --cflags" which often contains too much information and not just the basic information needed for MET_PYTHON_CC.

It sounds like you were able to get a successful compilation of MET.  Is that correct?  If not, please let us know.  If so, please feel free to select "Mark as answer". Our team has decided that once discussions have been answered, and the answer has been confirmed, we’ll lock them. We want to encourage users to ask new questions in new discussions rather than posting to old ones. Hopefully that’ll make the questions/answers easier for other users to follow. So if/when more issues/questions arise, please feel free to start a new discussion.

**Comment by HathewayWill** [comment]
Good afternoon Julie,

So running "python3-config --ldflags" shows tihs results on my system:

-L/opt/intel/oneapi/intelpython/python3.9/lib/python3.9/config-3.9-x86_64-linux-gnu
> -L/opt/intel/oneapi/intelpython/python3.9/lib  -lcrypt -lpthread -ldl
>  -lutil -lm -lm
>

So for some reason the -libutil is not needed when installing met on my
machine.  Not sure if anyone else has seen this similar issue.

For reference I am installing the intel compilers following this tutorial:

https://forum.mmm.ucar.edu/threads/how-to-install-intel-compilers-for-wrf.11904/#post-28523

**Reply by jprestop** [reply]
It looks like the -lpython3.9 is missing from what is needed for MET.  For example, it looks like this is what you should set MET_PYTHON_LD to:
```
-L/opt/intel/oneapi/intelpython/python3.9/lib/python3.9/config-3.9-x86_64-linux-gnu  -L/opt/intel/oneapi/intelpython/python3.9/lib -lpython3.9 -lcrypt -lpthread -ldl -lutil -lm
```
Although it sounds like you're saying the -lutil may not be needed, which is fine.  I haven't encountered not needing -lutil, but if your system doesn't seem to work with it, but will work if it is removed, go ahead and remove it and let us know if you encounter any difficultly.

**Comment by HathewayWill** [comment]
Sorry I meant -lutil not -libutil.

**Comment by HathewayWill** [comment, marked as answer]
Good evening,

MET_PYTHON_LD =
-L/opt/intel/oneapi/intelpython/python3.9/lib/python3.9/config-3.9-x86_64-linux-gnu
-L/opt/intel/oneapi/intelpython/python3.9/lib -lpython3.9 -lcrypt -lpthread
-ldl -lm -lm

Some reason -lutil doesn't need to be in for intel compilers. again not
sure why.  But it works for me and seems to be working for the test cases.

-Will

**Reply by jprestop** [reply]
@HathewayWill  Thank you for letting us know.  If you have no further questions in this discussion, please go ahead and pick the answer that you feel is the best one for this Discussion by selecting "Mark as answer" then I will lock this Discussion.
